clang/lib/CConv/ConstraintBuilder.cpp

```diff
@@ -173,10 +174,22 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
         }
         // and for each arg to the function ...
         if (FVConstraint *TargetFV = dyn_cast<FVConstraint>(TmpC)) {
+          std::set<const TypeVariableType *> InstantiableTypes;
+          if(TFD != nullptr)
+            getInstantiableTypeParams(E, TFD, InstantiableTypes);
+
           unsigned i = 0;
           for (const auto &A : E->arguments()) {
-            std::set<ConstraintVariable *> ArgumentConstraints =
-                CB.getExprConstraintVars(A);
+            std::set<ConstraintVariable *> ArgumentConstraints;
+            if(TFD != nullptr && i < TFD->getNumParams()) {
+              const auto *Ty = getTypeVariableType(TFD->getParamDecl(i));
+              if (InstantiableTypes.find(Ty) != InstantiableTypes.end())
+                ArgumentConstraints = CB.getExprConstraintVars(ignoreCast(A));
```

Comment:
Is it just this outermost implicit cast that's the issue? Can there be a sequence of implicit casts? I assume not; just curious if you know definitively (clang documentation) or it just seems to work in practice.

Reply:
I might add a comment around here that says you are just stripping off the cast if it's a polymorphic type; which type it is and whether it's instantiated properly are not your concerns.

Reply:
Having multiple implicit casts doesn't make much sense to me, but I don't see anything explicitly saying it can't happen.

Reply:
Maybe use this function rather than your home-grown one? https://clang.llvm.org/doxygen/classclang_1_1Expr.html#a99dd6d3939dbb82a3c8ced3b00dda30c

Reply:
Looks like it does the same thing. Thanks, nice to know that exists.